clients/web/src/hooks.ts

```diff
@@ -1,15 +1,94 @@
 import type {GetSession, Handle} from "@sveltejs/kit";
 import {constants, loadConfig} from "$lib/utils";
+import {add} from "date-fns";
+
+interface refreshPayload {
+    cookies: string[];
+    cookiesString: string;
+    accessToken: string;
+    refreshToken: string;
+}
+
+const doAuthRefresh = async (refreshUrl: string, currentCookies: string): Promise<refreshPayload> => {
+
+    let newCookiesString = "";
+    // If  refresh it
+    // Get the refresh token out of user's cookies
+    const refreshToken = currentCookies.split("; ").find(c => c.split("=")[0] === constants.refresh_token_cookie_key)
+        ?.split("=")[1];
+    if (refreshToken) {
+        // Send that refresh token to the backend, asking
+        // for a new JWT
+        const res = await fetch(refreshUrl, {
+            method: "GET",
+            headers: {
+                "Authorization": `Bearer ${refreshToken}`,
+            },
+        });
+
+        // Pull the new tokens out of the response
+        const tokens = await res.json();
+        const access_token = tokens.access_token;
+        const new_refresh_token = tokens.refresh_token;
+
+        // Store these new tokens back in the user's cookies
+        const newCookies = currentCookies.split("; ");
+
+        // Access token cookie
+        let newCookies1: string[];
+        if (newCookies.some(c => c.split("=")[0] === constants.auth_cookie_key)) {
+            // Cookie still exists, just update it
+            newCookies1 = newCookies.map(c => {
+                if (c.split("=")[0] === constants.auth_cookie_key) {
+                    return `${constants.auth_cookie_key}=${access_token};expires=${add(new Date(), {weeks: 1}).toUTCString()}`;
+                }
+                return c;
+            });
+        } else {
+            // Access token cookie doesn't exist, add it.
+            newCookies1 = newCookies;
+            newCookies1.push(`${constants.auth_cookie_key}=${access_token};expires=${add(new Date(), {weeks: 1}).toUTCString()}`);
+        }
+
+        // Refresh token cookie
+        const newCookies2 = newCookies1.map(c => {
+            if (c.split("=")[0] === constants.refresh_token_cookie_key) {
+                return `${constants.refresh_token_cookie_key}=${new_refresh_token};expires=${add(new Date(), {weeks: 1}).toUTCString()}`;
+            }
+            return c;
+        });
+
+        // Roll our cookies back up into one string
+        newCookiesString = newCookies2.join("; ");
+
+        return {
+            cookies: newCookies2,
+            cookiesString: newCookiesString,
+            accessToken: access_token,
+            refreshToken: new_refresh_token,
+        };
+    }
+
+    return {
+        cookies: [""],
+        cookiesString: "",
+        accessToken: "",
+        refreshToken: "",
```

Comment:
Do we know if this will wipe a user's cookies? If so, is that the behavior that we want?

Reply:
I don't think so, but I haven't tested it. Sending an empty "Set-cookies" HTTP request shouldn't wipe cookies, just not add any new ones.